### PR TITLE
Add uploading IP address info to cloud for switched net, add broadcas…

### PR DESCRIPTION
**Pull request by naiming-zededa**
…t mac for dhcp server in snooping

Signed-off-by: Naiming Shen <naiming@zededa.com>

Currently the NI status includes the switchnet ip address to be displayed in the cloud when in the NI basic-info page, but this information is more naturally needed in the App info section. Add the supprt in the app-info on device.

### CODE REVIEW

File: pkg/pillar/cmd/zedagent/handlemetrics.go [hunk 1]
```diff
@@ -1461,6 +1461,7 @@ func PublishAppInfoToZedCloud(ctx *zedagentContext, uuid string,
 				networkInfo := getNetInfo(interfaceDetail, false)
 				ip, allocated, macAddr := getAppIP(ctx, aiStatus,
 					ifname)
+				log.Infof("PublishAppInfoToZedCloud: after getAppIP, ifname %s, ip %s, mac %s", ifname, ip, macAddr)
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlemetrics.go]: This is in Handle metrics.. Getting sent repeatedly.. Why do u need it to be an Info message? Is it just needed only for debugging when developing? We anyway get to see the value in the Status sent to the cloud?

File: pkg/pillar/cmd/zedagent/handlemetrics.go [hunk 2]
```diff
@@ -1624,15 +1625,28 @@ func getDefaultRouters(ifname string) []string {
...
+		appIP := ulStatus.AllocatedIPAddr
+		ulconfig := ulStatus.UnderlayNetworkConfig
+		st, _ := sub.Get(ulconfig.Network.String())
+		if st != nil {
+			netstatus := cast.CastNetworkInstanceStatus(st)
+			if netstatus.Type == types.NetworkInstanceTypeSwitch {
+				log.Infof("getAppIP: type switch on %s", netstatus.BridgeName)
+				if _, ok := netstatus.IPAssignments[ulStatus.Mac]; ok {
+					appIP = netstatus.IPAssignments[ulStatus.Mac].String()
+					log.Infof("getAppIP: got sw net ip of app %s", appIP)
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlemetrics.go]: Same here.. do we really need this to be an Info msg? 

These periodic messages clutter the the debug log output..

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go [hunk 3]
```diff
@@ -761,6 +761,56 @@ func handleAppFlowMonitorDelete(ctxArg interface{}, key string,
 	log.Infof("handleAppFlowMonitorDelete(%s) done\n", key)
 }
 
+func handleAppVifIPTrigModify(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	log.Infof("handleAppVifIPTrigModify(%s)\n", key)
+	ctx := ctxArg.(*zedagentContext)
+	trig := cast.CastVifIPTrig(statusArg)
+	if trig.Key() != key {
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: You don't need this.. This was obsolete code. Please remove the If block.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go [hunk 4]
```diff
@@ -761,6 +761,56 @@ func handleAppFlowMonitorDelete(ctxArg interface{}, key string,
...
+		log.Errorf("handleAppVifIPTrigModify key mismatch %s vs %s; ignored\n",
+			key, trig.Key())
+		return
+	}
+	findVifAndTrigAppInfoUpload(ctx, trig.Key())
+}
+
+func handleAppVifIPTrigDelete(ctxArg interface{}, key string,
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: Is this a No-Op function? I don't see it deleting anything??

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go [hunk 5]
```diff
@@ -761,6 +761,56 @@ func handleAppFlowMonitorDelete(ctxArg interface{}, key string,
...
+	statusArg interface{}) {
+
+	log.Infof("handleAppVifIPTrigDelete(%s)\n", key)
+	trig := cast.CastVifIPTrig(statusArg)
+	if trig.Key() != key {
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: don;t need this if check.. Is Anything - it should be log.Fatalf

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go [hunk 6]
```diff
@@ -761,6 +761,56 @@ func handleAppFlowMonitorDelete(ctxArg interface{}, key string,
...
+}
+
+func findVifAndTrigAppInfoUpload(ctx *zedagentContext, macAddr string) {
+	sub := ctx.getconfigCtx.subAppInstanceStatus
+	items := sub.GetAll()
+
+	for _, st := range items {
+		var aiFound bool
+		aiStatus := cast.CastAppInstanceStatus(st)
+		for _, ulStatus := range aiStatus.UnderlayNetworks {
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: Is this If block better be a member function of aiStatus??
 
aiUsesMac := aiStatus.HasUnderlayNetworkWithMacAddr(macAddr)
 ...

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 7]
```diff
@@ -862,6 +873,12 @@ func Run() {
 			subAppFlowMonitor.ProcessChange(change)
 			agentlog.CheckMaxTime(agentName, start)
 
+		case change := <-subAppVifIPTrig.C:
+			start := agentlog.StartTime()
+			log.Debugf("VifIPTrig: change called")
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/zedagent.go]: Do we need this Debug statement??

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 1 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: sure. will remove this info

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 2 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: will do.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go (hunk 3 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: ok. was copying from someplace.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go (hunk 4 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: don't think we'll have a delete need on this call.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go (hunk 5 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: ok.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go (hunk 6 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlenetworkinstance.go]: make sense. will do.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 7 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: can remove.

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 2 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: BTW, this upload of info is only called once, that's why I need to have an explicit trigger for upload again;-)

**Comment by naiming-zededa** [comment]
@kalyan-nidumolu I'll put this adding the test case to my list...

File: pkg/pillar/types/zedmanagertypes.go [hunk 8]
```diff
@@ -198,6 +198,17 @@ func (statusPtr *AppInstanceStatus) ClearError() { //revive:disable-line
 	statusPtr.ErrorTime = time.Time{}
 }
 
+// HasUnderlayMacAddr - Check if the AI status has the underlay network with this Mac Address
+func (status *AppInstanceStatus) HasUnderlayMacAddr(MacAddr, ipAddr string) bool {
+	for idx, ulStatus := range status.UnderlayNetworks {
+		if ulStatus.VifInfo.Mac == MacAddr {
+			status.UnderlayNetworks[idx].AllocatedIPAddr = ipAddr
```
Comment (eriknordmark) [inline pkg/pillar/types/zedmanagertypes.go]: The function name reads as a predicate, which makes it confusing to have it also update the status.
Should it be named MaybeUpdateIPAddr() instead, or make the caller update the status?

Comment (naiming-zededa) [reply pkg/pillar/types/zedmanagertypes.go]: Sure.

**Pull request merged**